File: Node/utils/compare.images.js
const { S3Client, HeadObjectCommand,} = require("@aws-sdk/client-s3");
const { RekognitionClient, CompareFacesCommand } = require("@aws-sdk/client-rekognition");

const AWS = require('aws-sdk');
const { config } = require("dotenv");
require('dotenv').config();

const s3 = new S3Client({
    region: process.env.AWS_REGION,
    credentials: {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID,
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
    }
});

async function checkObjectExists(bucketName, objectKey) {
    try {
        const command = new HeadObjectCommand({ Bucket: bucketName, Key: objectKey });
        await s3.send(command);
        return true;
    } catch (error) {
        return false;
    }
}

const rekognition = new AWS.Rekognition({
    accessKeyId: process.env.REKOGNITION_ACCESS_KEY_ID,
    secretAccessKey: process.env.REKOGNITION_SECRET_ACCESS_KEY,
    region: process.env.REKOGNITION_REGION,
});

const config2 = {
    region: process.env.REKOGNITION_REGION,
    credentials: {
        accessKeyId: process.env.REKOGNITION_ACCESS_KEY_ID,
        secretAccessKey: process.env.REKOGNITION_SECRET_ACCESS_KEY
    }
};

const client = new RekognitionClient(config2);

async function compareImages(imageUrl, imageUrl2) {
    const imageName = imageUrl.split('/').pop();
    const imageName2 = imageUrl2.split('/').pop();

    const imageNameI1 = `Fotos_Reconocimiento_Facial/${imageName}`;
    const imageNameI2 = `Fotos_Reconocimiento_Facial/${imageName2}`;

    const existsI1 = await checkObjectExists(process.env.AWS_BUCKET_NAME, imageNameI1);
    const existsI2 = await checkObjectExists(process.env.AWS_BUCKET_NAME, imageNameI2);

    if (!existsI1 || !existsI2) {
        throw new Error('One or both images do not exist in the bucket.');
    }
    try {
        const params = {
            SimilarityThreshold: 80,
            SourceImage: {
                S3Object: {
                    Bucket: process.env.AWS_BUCKET_NAME,
                    Name: imageNameI1,
                },
            },
            TargetImage: {
                S3Object: {
                    Bucket: process.env.AWS_BUCKET_NAME,
                    Name: imageNameI2,
                },
            }
        };
        const result = await rekognition.compareFaces(params).promise();
        return result['FaceMatches'];
    } catch (error) {
        console.error('Error comparing images:', error);
        throw error;
    }
}

module.exports = compareImages;

File: Python/utils/compareImages.py
import boto3
import os
from dotenv import load_dotenv

load_dotenv()

# Verificar variables de entorno
print("AWS_REGION:", os.getenv('AWS_REGION'))
print("REKOGNITION_REGION:", os.getenv('REKOGNITION_REGION'))
print("AWS_ACCESS_KEY_ID:", os.getenv('AWS_ACCESS_KEY_ID'))  # Para verificar
print("AWS_SECRET_ACCESS_KEY:", os.getenv('AWS_SECRET_ACCESS_KEY'))  # Para verificar


rekognition = boto3.client('rekognition',
    aws_access_key_id=os.getenv('REKOGNITION_ACCESS_KEY_ID'),
    aws_secret_access_key=os.getenv('REKOGNITION_SECRET_ACCESS_KEY'),
    region_name=os.getenv('REKOGNITION_REGION_NAME')
)

def compare_images(image_url, image_url2):
    try:
        image_name = image_url.split('/').pop()
        image_name2 = image_url2.split('/').pop()

        print(rekognition.list_collections())
        
        params = {
            'SimilarityThreshold': 80,
            'SourceImage': {
                'S3Object': {
                    'Bucket': os.getenv('AWS_BUCKET_NAME'),
                    'Name': image_name,
                }
            },
            'TargetImage': {
                'S3Object': {
                    'Bucket': os.getenv('AWS_BUCKET_NAME'),
                    'Name': image_name2,
                }
            }
        }

        result = rekognition.compare_faces(**params)
        return result
    
    except Exception as e:
        print(f"Error al comparar las imágenes: {e}")
        raise e